### DISCUSSION TITLE
Is there a reason why `TryFromJsValue` takes ownership of the `JsValue`?

### DISCUSSION THREAD

**Discussion by hrxi** [wasm-bindgen/wasm-bindgen#4022] [Q&A] [answered]
The [`wasm-bindgen-derive`](https://docs.rs/wasm-bindgen-derive/0.3.0/wasm_bindgen_derive/) crate creates `impl TryFrom<&JsValue> for <type>`.

Would it be possible to take `&JsValue` in `TryFromJsValue`?

**Comment by daxpedda** [comment]
`wasm-bindgen-derive` is not maintained by us (`rustwasm` and `wasm-bindgen`).

Apologies for the very late response!

**Comment by hrxi** [comment]
The question wasn't about `wasm-bindgen-derive`. It's about [`wasm_bindgen::convert::TryFromJsValue`](https://docs.rs/wasm-bindgen/0.2.92/wasm_bindgen/convert/trait.TryFromJsValue.html). Could you re-open the issue?

**Comment by daxpedda** [comment]
Ah, I see!

So `TryFromJsValue` is an internal API, unless we intend to use it internally in `wasm-bindgen` somehow, there would be no reason to change that.

What exactly are you trying to achieve?
Maybe a public API is missing for what you are trying to do?

**Comment by hrxi** [comment]
> So `TryFromJsValue` is an internal API

I'm not sure what you mean by "internal API". It seems to be exposed: [`wasm_bindgen::convert::TryFromJsValue`](https://docs.rs/wasm-bindgen/0.2.92/wasm_bindgen/convert/trait.TryFromJsValue.html).

We considered using it, but it requires cloning to turn the `&JsValue` into a `JsValue`. With `wasm-bindgen-derive`, we could do the following (`js_value_ref` is of type `&JsValue`)
```rust
PublicKey::try_from(js_value_ref)
```
With `TryFromJsValue`, we need to do
```rust
PublicKey::try_from_js_value(js_value_ref.to_owned())
```

Could the `TryFromJsValue` trait be changed to take the incoming `JsValue` by reference instead of requiring an owned copy?

**Comment by daxpedda** [comment, marked as answer]
> > So `TryFromJsValue` is an internal API
> 
> I'm not sure what you mean by "internal API". It seems to be exposed: [`wasm_bindgen::convert::TryFromJsValue`](https://docs.rs/wasm-bindgen/0.2.92/wasm_bindgen/convert/trait.TryFromJsValue.html).

It says right there at the top of the module:
> This is mostly an internal module, no stability guarantees are provided. Use at your own risk.

Unless you know exactly what you are doing, don't use this API.
Its meant for proc-macros and using `wasm-bindgen` descriptors to define how exports/imports are generated.

---

What kind of type is `PublicKey`?
I assume its an exported type?
You should probably take a look at [`JsCast`](https://docs.rs/wasm-bindgen/0.2.92/wasm_bindgen/trait.JsCast.html).

**Reply by hrxi** [reply]
> It says right there at the top of the module:
> 
> > This is mostly an internal module, no stability guarantees are provided. Use at your own risk.

Thanks. That clarifies that we shouldn't use it. Thanks.

**Reply by sisou** [reply]
@hrxi To be fair, the external crate says in it's documentation that the type it is derived for must implement `Clone` and it uses it, too: https://docs.rs/wasm-bindgen-derive-macro/0.3.0/src/wasm_bindgen_derive_macro/lib.rs.html#171

So basically the external crate calls the `.clone()` for us, whereas when using the internal trait, we explicitly clone it ourselves.
